# enable gpu build, remove after next pin update
diff --git a/tensorflow/core/kernels/fft_ops.cc b/tensorflow/core/kernels/fft_ops.cc
index 5e0464e89f1..2d9cd438eb2 100644
--- a/tensorflow/core/kernels/fft_ops.cc
+++ b/tensorflow/core/kernels/fft_ops.cc
@@ -455,7 +455,7 @@ class FftPlanCache {
       cache_.erase(it);
     }
     --size_;
-    return value;
+    return std::optional<Value>(std::move(value));
   }
 
   // Inserts a plan into the cache as long as there is still capacity.
