https://github.com/pytorch/xla/issues/4084
diff --git a/tensorflow/core/distributed_runtime/request_id.cc b/tensorflow/core/distributed_runtime/request_id.cc
index 24fc3004391..4a8954072c8 100644
--- a/tensorflow/core/distributed_runtime/request_id.cc
+++ b/tensorflow/core/distributed_runtime/request_id.cc
@@ -22,7 +22,7 @@ namespace tensorflow {
 int64_t GetUniqueRequestId() {
   int64_t request_id = 0;
   while (request_id == 0) {
-    request_id = tsl::random::ThreadLocalNew64();
+    request_id = tsl::random::New64();
   }
   return request_id;
 }
diff --git a/tensorflow/core/distributed_runtime/request_id.h b/tensorflow/core/distributed_runtime/request_id.h
index 1ef60c8f47c..c597a15fcc9 100644
--- a/tensorflow/core/distributed_runtime/request_id.h
+++ b/tensorflow/core/distributed_runtime/request_id.h
@@ -16,8 +16,8 @@ limitations under the License.
 #ifndef TENSORFLOW_CORE_DISTRIBUTED_RUNTIME_REQUEST_ID_H_
 #define TENSORFLOW_CORE_DISTRIBUTED_RUNTIME_REQUEST_ID_H_
 
+#include "tensorflow/core/lib/random/random.h"
 #include "tensorflow/core/platform/types.h"
-#include "tensorflow/tsl/platform/random.h"
 
 namespace tensorflow {
diff --git a/tensorflow/tsl/platform/random.cc b/tensorflow/tsl/platform/random.cc
index 636725ad650..02d0ecda23d 100644
--- a/tensorflow/tsl/platform/random.cc
+++ b/tensorflow/tsl/platform/random.cc
@@ -15,7 +15,6 @@ limitations under the License.
 
 #include "tensorflow/tsl/platform/random.h"
 
-#include <memory>
 #include <random>
 
 #include "tensorflow/tsl/platform/mutex.h"
@@ -40,12 +39,6 @@ uint64 New64() {
   return (*rng)();
 }
 
-uint64 ThreadLocalNew64() {
-  static thread_local std::unique_ptr<std::mt19937_64> rng =
-      std::unique_ptr<std::mt19937_64>(InitRngWithRandomSeed());
-  return (*rng)();
-}
-
 uint64 New64DefaultSeed() {
   static std::mt19937_64 rng = InitRngWithDefaultSeed();
   static mutex mu(LINKER_INITIALIZED);
diff --git a/tensorflow/tsl/platform/random.h b/tensorflow/tsl/platform/random.h
index 9e08670a2c6..e13555c5128 100644
--- a/tensorflow/tsl/platform/random.h
+++ b/tensorflow/tsl/platform/random.h
@@ -25,9 +25,6 @@ namespace random {
 // in different processes.
 uint64 New64();
 
-// Same as previous method, but uses a different RNG for each thread.
-uint64 ThreadLocalNew64();
-
 // Return a 64-bit random value. Uses
 // std::mersenne_twister_engine::default_seed as seed value.
 uint64 New64DefaultSeed();

