diff --git a/xla/pjrt/gpu/BUILD b/xla/pjrt/gpu/BUILD
index 9ad1fca31..8ea07ad0a 100644
--- a/xla/pjrt/gpu/BUILD
+++ b/xla/pjrt/gpu/BUILD
@@ -237,17 +237,17 @@ cc_library(
         "@com_google_absl//absl/status",
         "@tsl//tsl/platform:casts",
         "@tsl//tsl/platform:errors",
-    ] + if_cuda([
+    ] + if_cuda_or_rocm([
+        "//xla/service/gpu:gpu_compiler",
+    ]) + if_cuda([
         ":nccl_id_store_cuda",
         "@local_config_cuda//cuda:cuda_headers",
         "//xla/stream_executor/cuda:cuda_activation_header",
         "//xla/stream_executor/gpu:gpu_cudamallocasync_allocator",
-        "//xla/service/gpu:gpu_compiler",
         "//xla/service/gpu:nvptx_compiler_impl",
     ]) + if_rocm([
         ":nccl_id_store_rocm",
         "@local_config_rocm//rocm:rocm_headers",
-        "//xla/service/gpu:gpu_compiler",
         "//xla/service/gpu:amdgpu_compiler_impl",
     ]),
     alwayslink = True,
