diff --git a/xla/backends/profiler/gpu/cupti_error_manager.cc b/xla/backends/profiler/gpu/cupti_error_manager.cc
index 5a7cd7f52a..a436e211e9 100644
--- a/xla/backends/profiler/gpu/cupti_error_manager.cc
+++ b/xla/backends/profiler/gpu/cupti_error_manager.cc
@@ -167,6 +167,10 @@ CUptiResult CuptiErrorManager::EnableCallback(uint32_t enable,
                          0 /* DISABLE */, subscriber, domain, callback_id);
       RegisterUndoFunction(f);
     }
+  } else {
+    LOG(ERROR) << "cupti" << __func__
+               << ": error with domain:" << static_cast<int>(domain)
+               << " and callback_id:" << static_cast<int>(callback_id);
   }
   LOG_AND_DISABLE_IF_ERROR(error);
   return error;
diff --git a/xla/backends/profiler/gpu/cupti_tracer.cc b/xla/backends/profiler/gpu/cupti_tracer.cc
index f2119ddc74..50897f0851 100644
--- a/xla/backends/profiler/gpu/cupti_tracer.cc
+++ b/xla/backends/profiler/gpu/cupti_tracer.cc
@@ -932,7 +932,7 @@ class CuptiDriverApiHookWithActivityApi : public CuptiDriverApiHook {
 
 absl::Span<const uint32_t> GetCudaGraphTracingResourceCbids() {
 #if CUDA_VERSION >= 11070
-  constexpr uint32_t res_cbids[] = {CUPTI_CBID_RESOURCE_GRAPH_CREATED,
+  static constexpr uint32_t res_cbids[] = {CUPTI_CBID_RESOURCE_GRAPH_CREATED,
                                     CUPTI_CBID_RESOURCE_GRAPH_CLONED,
                                     CUPTI_CBID_RESOURCE_GRAPHEXEC_CREATED};
   return absl::MakeSpan(res_cbids);
