diff --git a/xla/stream_executor/BUILD b/xla/stream_executor/BUILD
index 7d56354f7..24545a7ef 100644
--- a/xla/stream_executor/BUILD
+++ b/xla/stream_executor/BUILD
@@ -212,10 +212,12 @@ cc_library(
     ],
 )
 
+# We need to remove make_default_target_header_only so that
+# we can build PyTorch/XLA. Otherwise, we won't have the
+# required objects linked at build time.
 tf_proto_library(
     name = "dnn_proto",
     srcs = ["dnn.proto"],
-    make_default_target_header_only = True,
     protodeps = ["@tsl//tsl/protobuf:dnn_proto"],
     exports = ["@tsl//tsl/protobuf:dnn_proto"],
 )