commit 55fbb131fd1676365af72417aeb46292251075f3
Author: Davide Libenzi <dlibenzi@google.com>
Date:   Thu Feb 7 08:35:20 2019 -0800

    Add recognition for XLA device types.

diff --git a/c10/core/Device.cpp b/c10/core/Device.cpp
index 1d2d1ec0b..79ee6251e 100644
--- a/c10/core/Device.cpp
+++ b/c10/core/Device.cpp
@@ -13,7 +13,7 @@
 namespace c10 {
 namespace {
 DeviceType parse_type(const std::string& device_string) {
-  static const std::array<std::pair<std::string, DeviceType>, 8> types = {{
+  static const std::array<std::pair<std::string, DeviceType>, 9> types = {{
       {"cpu", DeviceType::CPU},
       {"cuda", DeviceType::CUDA},
       {"mkldnn", DeviceType::MKLDNN},
@@ -22,6 +22,7 @@ DeviceType parse_type(const std::string& device_string) {
       {"ideep", DeviceType::IDEEP},
       {"hip", DeviceType::HIP},
       {"msnpu", DeviceType::MSNPU},
+      {"xla", DeviceType::XLA},
   }};
   auto device = std::find_if(
       types.begin(),
