commit 4502becf2d584a334f764b3bf388abc191660e4d
Author: Davide Libenzi <dlibenzi@google.com>
Date:   Sun Feb 10 18:18:50 2019 -0800

    Register layout for XLA backend.

diff --git a/torch/csrc/utils/tensor_layouts.cpp b/torch/csrc/utils/tensor_layouts.cpp
index 7d984f52d..4b1b36f46 100644
--- a/torch/csrc/utils/tensor_layouts.cpp
+++ b/torch/csrc/utils/tensor_layouts.cpp
@@ -24,6 +24,7 @@ void initializeLayouts() {
   registerLayoutObject((THPLayout*)strided_layout, at::Backend::CPU);
   registerLayoutObject((THPLayout*)strided_layout, at::Backend::CUDA);
   registerLayoutObject((THPLayout*)strided_layout, at::Backend::MSNPU);
+  registerLayoutObject((THPLayout*)strided_layout, at::Backend::XLA);
 
   PyObject *sparse_coo_layout = THPLayout_New(at::Layout::Sparse, "torch.sparse_coo");
   Py_INCREF(sparse_coo_layout);
