commit e3c6267429670a6935d50050289b5819b536b488
Author: Alex Şuhan <asuhan@google.com>
Date:   Wed Feb 20 19:13:45 2019 -0800

    Fix Python device type property for XLA and MSNPU

diff --git a/torch/csrc/tensor/python_tensor.cpp b/torch/csrc/tensor/python_tensor.cpp
index 54bda656d..b833d4bbe 100644
--- a/torch/csrc/tensor/python_tensor.cpp
+++ b/torch/csrc/tensor/python_tensor.cpp
@@ -392,6 +392,9 @@ Device getDevice(const at::Tensor& tensor) {
   if (tensor.is_cuda()) {
     return at::Device(at::DeviceType::CUDA, tensor.get_device());
   }
+  if (tensor.type_id() == XLATensorId() || tensor.type_id() == MSNPUTensorId()) {
+    return at::Device(computeDeviceType(tensor.type_id()), tensor.get_device());
+  }
   return at::Device(at::DeviceType::CPU);
 }
 }} // namespace torch::tensors
