diff --git a/torch/serialization.py b/torch/serialization.py
index 98993301b9..05bc139813 100644
--- a/torch/serialization.py
+++ b/torch/serialization.py
@@ -362,6 +362,9 @@ def save(obj, f, pickle_module=pickle, pickle_protocol=DEFAULT_PROTOCOL, _use_ne
     """
     _check_dill_version(pickle_module)
 
+    import torch_xla.core.xla_model as xm
+    xm.check_view_sharing(obj)
+
     with _open_file_like(f, 'wb') as opened_file:
         if _use_new_zipfile_serialization:
             with _open_zipfile_writer(opened_file) as opened_zipfile:
