diff --git a/torch/serialization.py b/torch/serialization.py
index b38422a7fc..3d601422e5 100644
--- a/torch/serialization.py
+++ b/torch/serialization.py
@@ -267,6 +267,8 @@ def save(obj, f, pickle_module=pickle, pickle_protocol=DEFAULT_PROTOCOL):
         >>> buffer = io.BytesIO()
         >>> torch.save(x, buffer)
     """
+    import torch_xla_py.xla_model as xm
+    xm.check_view_sharing(obj)
     return _with_file_like(f, "wb", lambda f: _save(obj, f, pickle_module, pickle_protocol))
 
 
