diff --git a/torch/serialization.py b/torch/serialization.py
index 1c1c5f21b0..80748971fa 100644
--- a/torch/serialization.py
+++ b/torch/serialization.py
@@ -319,6 +319,8 @@ def save(obj, f, pickle_module=pickle, pickle_protocol=DEFAULT_PROTOCOL, _use_ne
         >>> buffer = io.BytesIO()
         >>> torch.save(x, buffer)
     """
+    import torch_xla.core.xla_model as xm
+    xm.check_view_sharing(obj)
     if _use_new_zipfile_serialization:
         with _open_zipfile_writer(f) as opened_file:
             _save(obj, opened_file, pickle_module, pickle_protocol)
