diff --git a/torch/tensor.py b/torch/tensor.py
index 0c3286078b..cf0752c89a 100644
--- a/torch/tensor.py
+++ b/torch/tensor.py
@@ -38,6 +38,8 @@ class Tensor(torch._C._TensorBase):
     def __reduce_ex__(self, proto):
         # See Note [Don't serialize hooks]
         torch.utils.hooks.warn_if_has_hooks(self)
+        if self.device.type == 'xla':
+            self = self.cpu()
         if self.is_quantized:
             args = (self.storage(),
                     self.storage_offset(),
